### PR TITLE
Authentication improvements 

### PR DESCRIPTION
**Pull request by Isakdl**
# Changes
- Restructure the authentication documentation into several pages.
- Restructures each provider in the auth module into their own pages.
- Expand each provider with more detailed instructions for the setup.

## Closes
https://github.com/serverpod/serverpod/issues/843
https://github.com/serverpod/serverpod/issues/857
https://github.com/serverpod/serverpod/issues/544

## Todo
https://github.com/serverpod/serverpod/issues/350

### CODE REVIEW

**Comment by easazade** [comment]
@Isakdl I like a detailed documentation. 👍

File: docs/04-concepts/09-authentication/03-providers/02-google.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+To be allowed to access user data and use the authentication method in Serverpod we have to enable the Peoples API in our project.
+
+[Enable it here](https://console.cloud.google.com/apis/library/people.googleapis.com) or find it yourself by, navigating to the _Library_ section under _APIs & Services_. Search for _Google People API_, select it, and click on _Enable_.
+
+### Setup OAuth consent screen
+
+Can be found [here](https://console.cloud.google.com/apis/credentials/consent) or under _APIs & Services_ > _OAuth consent screen_.
+
+1. Fill in all the required information, for production use you need a domain that adds under `Authorized` domains`.
```
Comment (vlidholt) [inline docs/04-concepts/09-authentication/03-providers/02-google.md:29]: The trailing quote can mess up the formatting here.

File: docs/04-concepts/09-authentication/03-providers/03-apple.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# Apple
+
+Sign-in with Apple, requires that you have a subscription to the [Apple developer program](https://developer.apple.com/programs/), even if you only want to test the feature in development mode.
+
+A comprehensive tutorial covering Sign in with Apple is available [here](#todo).
```
Comment (vlidholt) [inline docs/04-concepts/09-authentication/03-providers/03-apple.md:5]: Let's upload it :)

File: docs/04-concepts/09-authentication/03-providers/03-apple.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+:::note
+Right now, we only have official support for iOS and MacOS for Sign in with Apple.
```
Comment (vlidholt) [inline docs/04-concepts/09-authentication/03-providers/03-apple.md:8]: Remove only. :P

**Review by vlidholt — CHANGES_REQUESTED** [review]
Great! Found some minor things. :)

File: docs/04-concepts/09-authentication/03-providers/01-email.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  },
+));
+
+// Start the Serverpod server.
+await pod.start();
+```
+
+:::info
+
+For debugging purposes, you can print the validation code to the console. The chat module example does just this. You can view that code [here](https://github.com/serverpod/serverpod/blob/main/examples/chat/chat_server/lib/server.dart).
```
[the remark this answers is not in the document]
Comment (Isakdl) [reply docs/04-concepts/09-authentication/03-providers/01-email.md:38]: Yes, the chat module still does this!

File: docs/04-concepts/09-authentication/01-the-basics.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+### Custom scopes
+
+You may need more granular access control for specific endpoints. To create custom scopes, extend the Scope class, as shown below:
+
+```dart
+class CustomScope extends Scope {
+  const CustomScope(String name) : super(name);
+
+  static const readUsers = CustomScope('read:users');
```
[the remark this answers is not in the document]
Comment (Isakdl) [reply docs/04-concepts/09-authentication/01-the-basics.md:74]: Not following what you mean, Scopes are just strings right or do we parse them in some special way?

Using `:` is quite common in scopes.

For example, slack:
https://api.slack.com/scopes

Could also be modeled with a `.` but the order should probably be reversed though. `resource:access_type` or `resource.access_type` I.E. `user.read`

**Comment by Isakdl** [comment]
@vlidholt check the update! :)

**Comment by braj065** [comment]
Will the changes in documentation contain details on how to authenticate(google api, firebase etc) using serverpod api calls directly and not using the pre-made serverpod auth widgets? And try to keep the serverpod auth api use in the docs higher/more preferred than the premade auth widgets.

**Comment by andriyndev** [comment]
It looks like info about error handling is missing.

**Pull request merged**